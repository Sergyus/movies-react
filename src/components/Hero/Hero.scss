@import "../../styles/variables.scss";

.hero {
  box-sizing: border-box;
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
  padding-bottom: 20px;
  height: 80vh;
  position: relative;
  background: no-repeat center;
  background-size: cover;

  &::after {
    content: "";
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    background: linear-gradient(0deg, black, transparent 40%);
  }
}

.title {
  font-size: 2rem;
  font-family: "Helvetica Neue", Arial, Helvetica, sans-serif;
  font-weight: bold;
  margin-bottom: 2%;
}

.duration {
  border-left: 1px solid white;
  padding-left: 3%;
}

.wrapper {
  margin-bottom: 20px;
  font-size: 0.9rem;
  font-family: "Helvetica Neue", Arial, Helvetica, sans-serif;
}

.movieInfo {
  width: 300px;
  color: white;
  margin-left: 5%;
  margin-bottom: 5px;
  z-index: 1;
}

.rating {
  font-size: 0.8rem;
  display: flex;
  align-items: flex-end;

  span {
    margin-left: 3%;
  }
}

.movieDescription {
  position: relative;
  z-index: 1;
  margin-right: 1%;
}

.description {
  opacity: 1;
  position: relative;
  width: 250px;
  height: 120px;
  border-radius: 3px;
  background-color: rgba(0, 0, 0, 0.6);
  color: white;
  padding: 20px;
  margin-bottom: 35px;
  font-size: 0.9rem;
  line-height: 1.3rem;

  &::after {
    content: '';
    border-left: 15px solid transparent;
    border-right: 15px solid transparent;
    border-top: 20px solid black;
    opacity: 0.6;
    position: absolute;
    right: 35%;
    bottom: -20px;
  }
}

.hiddenDesc {
  opacity: 0;
  width: 250px;
  height: 120px;
  line-height: 1.3rem;
  padding: 20px;
  margin-bottom: 35px;
}

.buttons {
  display: flex;
  z-index: 1;

  & :first-child {
    margin-right: 20px;
  }
}

@media (max-width: $tablet-screen-size) {
  .hero {
    flex-wrap: wrap;
  }

  .movieInfo {
    width: 150px;
    display: flex;
    flex-direction: column;
    align-items: center;

    & span:not(:last-child) {
      margin-bottom: 5px;
    }
  }

  .duration {
    border: none;
  }

  .wrapper {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
  }

  .description {
    &::after {
      display: none;
    }
  }

  .buttons {
    justify-content: center;
  }
}

@media (max-width: $mobile-screen-size) {
  .hero {
    justify-content: center;
    align-items: center;
  }

  .movieInfo {
    margin-left: 0;
    margin-top: 17vh;
    width: 90vw;
    display: flex;
    flex-direction: column;
    align-items: center;
    border-radius: 3px;
    background-color: rgba(0, 0, 0, 0.6);
    padding: 5px 0;

    & span:not(:last-child) {
      margin-bottom: 5px;
    }
  }

  .description {
    box-sizing: border-box;
    width: 90vw;
    height: auto;
  }

  .hiddenDesc {
    box-sizing: border-box;
    width: 90vw;
    height: auto;
  }

  .wrapper {
    display: block;
    width: 70vw;
    margin-bottom: 10px;
  }
}
