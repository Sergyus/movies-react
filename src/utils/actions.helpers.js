export const createAsyncFnRequest = requestName => ({
  type: requestName,
});

export const createAsyncFnFailure = (requestName, error) => ({
  type: requestName,
  error: error.message,
});
