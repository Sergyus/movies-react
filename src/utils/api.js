const MOVIE_DB_API_KEY = "ed98d25919802b5020e8feb6c3923886";
const MOVIE_DB_BASE_URL = "https://api.themoviedb.org/3";
const LANGUAGE = "ru-RU";

const createMovieDbUrl = (relativeUrl, queryParams) => {
  let baseUrl = `${MOVIE_DB_BASE_URL}${relativeUrl}?api_key=${MOVIE_DB_API_KEY}&language=${LANGUAGE}`;
  if (queryParams) {
    Object.keys(queryParams).forEach(paramName => {
      baseUrl += `&${paramName}=${queryParams[paramName]}`;
      return baseUrl;
    });
  }
  return baseUrl;
};

const getLatestMovies = ({ page }) => {
  const fullUrl = createMovieDbUrl("/movie/popular", {
    page
  });
  return fetch(fullUrl);
};

const getTopMovies = ({ page }) => {
  const fullUrl = createMovieDbUrl("/movie/top_rated", {
    page
  });
  return fetch(fullUrl);
};

const getUpcomingMovies = ({ page }) => {
  const fullUrl = createMovieDbUrl("/movie/upcoming", {
    page
  });
  return fetch(fullUrl);
};

const searchMovies = ({ page, query }) => {
  const fullUrl = `${createMovieDbUrl("/search/movie", {
    page,
    query
  })}&include_adult=false`;
  return fetch(fullUrl);
};

const getGenres = () => {
  const fullUrl = createMovieDbUrl("/genre/movie/list");
  return fetch(fullUrl);
};

const getMoviesByGenre = ({ page, genre }) => {
  const fullUrl = `${createMovieDbUrl("/discover/movie", {
    page
  })}&sort_by=popularity.desc&include_adult=false&include_video=false&with_genres=${genre}`;
  return fetch(fullUrl);
};

const getMovieDetails = ({ movieId }) => {
  const fullUrl = createMovieDbUrl(`/movie/${movieId}`);
  return fetch(fullUrl);
};

const getMovieVideo = movieId => {
  const fullUrl = createMovieDbUrl(`/movie/${movieId}/videos`);
  return fetch(fullUrl);
};

export {
  getTopMovies,
  searchMovies,
  getMovieDetails,
  getGenres,
  getMoviesByGenre,
  getLatestMovies,
  getUpcomingMovies,
  getMovieVideo
};
