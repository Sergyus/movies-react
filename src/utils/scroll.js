const getScrollDownPercentage = ({ scrollHeight, scrollPos, clientHeight }) => {
  const currentPosition = scrollPos + clientHeight;
  const percentageScrolled = currentPosition / scrollHeight;
  return percentageScrolled;
};

export default getScrollDownPercentage;
