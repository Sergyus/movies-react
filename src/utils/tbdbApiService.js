import noAvailableImage from '../assets/images/noAvailableImage.png';

const urlBase = 'https://image.tmdb.org/t/p/original';
const urlMinImg = 'https://image.tmdb.org/t/p/w300';

function getNormalRuntime(time) {
  const hours = Math.floor(time / 60);
  const minutes = time - 60 * hours;
  return `${hours}h ${minutes}m`;
}

export function getMovieGenres(genresIds = [], genres = []) {
  const genresArr = [];
  genresIds.map(genre => genres.map((movie) => {
    if (movie.id === genre) {
      return genresArr.push({ id: movie.id, name: movie.name });
    }
    return 0;
  }));
  return genresArr;
}

export function getMovieInfo(movie) {
  return {
    title: movie.title,
    genres: movie.genres,
    rating: movie.vote_average,
    runtime: getNormalRuntime(movie.runtime),
    overview: movie.overview,
    image: movie.backdrop_path ? `${urlBase}${movie.backdrop_path}` : noAvailableImage,
  };
}

export function getMovies(movies = [], allGenres = []) {
  return movies.map(movie => ({
    id: movie.id,
    rating: movie.vote_average,
    title: movie.title,
    genres: getMovieGenres(movie.genre_ids, allGenres),
    overview: movie.overview,
    image: movie.poster_path ? `${urlMinImg}${movie.poster_path}` : noAvailableImage,
  }));
}
