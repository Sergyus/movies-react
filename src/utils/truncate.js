const truncate = (description = '', maxLength) => (
  description.length && description.length > maxLength ? `${description.slice(0, maxLength)}...` : description
);

export default truncate;
