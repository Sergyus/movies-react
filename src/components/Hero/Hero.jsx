import React, { useState, useCallback } from 'react';
import cn from 'classnames/bind';
import styles from './Hero.scss';
import StarRating from '../StarRating';
import { getMovieInfo } from '../../utils/tbdbApiService';
import NumRating from '../NumRating';
import Genres from '../Genres';
import Button from '../Button';
import movieInfo from '../../assets/json/movie.json';

const cx = cn.bind(styles);
const Hero = () => {
  const [showInfo, setShowInfo] = useState(true);
  const movie = getMovieInfo(movieInfo);
  const {
    title,
    genres,
    rating,
    runtime,
    overview,
    image,
  } = movie;
  return (
    <section className={styles.hero} style={{ backgroundImage: `url(${image})` }}>
      <div className={styles.movieInfo}>
        <h2 className={styles.title}>{ title }</h2>
        <div className={styles.wrapper}>
          <Genres genres={genres} />
          <span className={styles.duration}>{runtime}</span>
        </div>
        <div className={styles.rating}>
          <StarRating
            starsSelected={rating}
          />
          <NumRating rating={rating} />
        </div>
      </div>
      <div className={styles.movieDescription}>
        <p className={cx({ description: showInfo, hiddenDesc: !showInfo })}>
          {`${overview.slice(0, 221)} ...`}
        </p>
        <div className={styles.buttons}>
          <Button type="primary">Watch Now</Button>
          <Button type="ghost" handleClick={useCallback(() => setShowInfo(!showInfo), [showInfo])}>View Info</Button>
        </div>
      </div>
    </section>
  );
};

export default Hero;
