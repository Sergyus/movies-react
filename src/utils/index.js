import {
  getTopMovies, searchMovies, getGenres, getMoviesByGenre,
  getLatestMovies, getUpcomingMovies, getMovieVideo,
} from './api';
import { getMovies } from './tbdbApiService';
import truncate from './truncate';
import { useFetchOnScroll } from './useFetchOnScroll';
import useBodyScrollLock from './useBodyScrollLock';
import useOnClickOutside from './useOnClickOutside';
import { createAsyncFnFailure, createAsyncFnRequest } from './actions.helpers';

export {
  getTopMovies, searchMovies, getGenres, getMoviesByGenre,
  getLatestMovies, getUpcomingMovies, getMovieVideo,
  truncate, useBodyScrollLock, useOnClickOutside, getMovies,
  createAsyncFnFailure, createAsyncFnRequest, useFetchOnScroll,
};
