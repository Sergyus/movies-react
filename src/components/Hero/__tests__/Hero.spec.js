import React from 'react';
import { create, act } from 'react-test-renderer';
import Hero from '..';
import { getMovieInfo } from '../../../utils/tbdbApiService';
import mockedMovie from '../../../assets/json/movie.json';

describe('Hero component renders correctly', () => {
  const movie = getMovieInfo(mockedMovie);

  it('Hero renders correctly', () => {
    const tree = create(<Hero movie={movie} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Hide description works correctly', () => {
    const tree = create(<Hero movie={movie} />);
    const button = tree.root.findByProps({ 'aria-label': 'View Info' });
    act(() => {
      button.props.onClick();
    });
    const description = tree.toJSON().children[1].children[0];
    expect(description.props.className).toEqual('hiddenDesc');
  });
});
