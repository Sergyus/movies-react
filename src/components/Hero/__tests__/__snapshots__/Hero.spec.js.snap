// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hero component renders correctly Hero renders correctly 1`] = `
<section
  className="hero"
  style={
    Object {
      "backgroundImage": "url(https://image.tmdb.org/t/p/original/59qyrVGLLtrcuFtJGdixzn7H4OI.jpg)",
    }
  }
>
  <div
    className="movieInfo"
  >
    <h2
      className="title"
    >
      T-34
    </h2>
    <div
      className="wrapper"
    >
      <span
        className="genre"
      >
        War
      </span>
      <span
        className="genre"
      >
        Drama
      </span>
      <span
        className="genre"
      >
        Adventure
      </span>
      <span
        className="genre"
      >
        Action
      </span>
      <span
        className="duration"
      >
        2h 19m
      </span>
    </div>
    <div
      className="rating"
    >
      <div
        className="star"
      />
      <div
        className="star"
      />
      <div
        className="star"
      />
      <div
        className="star"
      />
      <div
        className="star"
      />
      <span
        className="numRating"
      >
        4.7
      </span>
    </div>
  </div>
  <div
    className="movieDescription"
  >
    <p
      className="description"
    >
      1941 - WWii. the second lieutenant Nikolai ivushkin, commander of a t-34, engages in an unequal battle against the tank ace Klaus Jager in a battle near moscow. His mission is more of a suicide - to destroy a dozen german ...
    </p>
    <div
      className="buttons"
    >
      <button
        aria-label="Watch Now"
        className="btn primary"
        onClick={[Function]}
        type="button"
      >
        Watch Now
      </button>
      <button
        aria-label="View Info"
        className="btn ghost"
        onClick={[Function]}
        type="button"
      >
        View Info
      </button>
    </div>
  </div>
</section>
`;
